lib/cli.ts

```diff
@@ -31,49 +31,82 @@ process.argv.slice(2).forEach(function(arg: string) {
-    .alias('platform', 'capabilities.platform')
-    .alias('platform-version', 'capabilities.version')
-    .alias('tags', 'capabilities.tags')
-    .alias('build', 'capabilities.build')
-    .alias('grep', 'jasmineNodeOpts.grep')
-    .alias('invert-grep', 'jasmineNodeOpts.invertGrep')
-    .alias('explorer', 'elementExplorer')
-    .string('capabilities.tunnel-identifier')
-    .check(function(arg: any) {
-      if (arg._.length > 1) {
-        throw new Error('Error: more than one config file specified');
-      }
-    });
+let optimistOptions: any = {
+  describes: {
+    help: 'Print Protractor help menu',
+    version: 'Print Protractor version',
+    browser: 'Browsername, e.g. chrome or firefox',
+    seleniumAddress: 'A running selenium address to use',
+    seleniumSessionId: 'Attaching an existing session id',
+    seleniumServerJar: 'Location of the standalone selenium jar file',
+    seleniumPort: 'Optional port for the selenium standalone server',
+    baseUrl: 'URL to prepend to all relative paths',
+    rootElement: 'Element housing ng-app, if not html or body',
+    specs: 'Comma-separated list of files to test',
+    exclude: 'Comma-separated list of files to exclude',
+    verbose: 'Print full spec names',
+    stackTrace: 'Print stack trace on error',
+    params: 'Param object to be passed to the tests',
+    framework: 'Test framework to use: jasmine, mocha, or custom',
+    resultJsonOutputFile: 'Path to save JSON test result',
+    troubleshoot: 'Turn on troubleshooting output',
+    elementExplorer: 'Interactively test Protractor commands',
+    debuggerServerPort: 'Start a debugger server at specified port instead of repl'
+  },
+  aliases: {
+    browser: 'capabilities.browserName',
+    name: 'capabilities.name',
+    platform: 'capabilities.platform',
+    'platform-version': 'capabilities.version',
+    tags: 'capabilities.tags',
+    build: 'capabilities.build',
+    grep: 'jasmineNodeOpts.grep',
+    'invert-grep': 'jasmineNodeOpts.invertGrep',
+    explorer: 'elementExplorer'
+  },
+  strings: {'capabilities.tunnel-identifier': ''}
+};
+
+optimist.usage(
+    'Usage: protractor [configFile] [options]\n' +
+    'configFile defaults to protractor.conf.js\n' +
+    'The [options] object will override values from the config file.\n' +
+    'See the reference config for a full list of options.');
+for (let key of Object.keys(optimistOptions.describes)) {
+  optimist.describe(key, optimistOptions.describes[key]);
+}
+for (let key of Object.keys(optimistOptions.aliases)) {
+  optimist.alias(key, optimistOptions.aliases[key]);
+}
+for (let key of Object.keys(optimistOptions.strings)) {
+  optimist.string(key);
+}
+optimist.check(function(arg: any) {
+  if (arg._.length > 1) {
+    throw new Error('Error: more than one config file specified');
+  }
+});
 
 let argv: any = optimist.parse(args);
 
+// Check to see if additional flags were used.
+let unknownKeys: string[] = [];
+for (let key of Object.keys(argv).slice(2)) {
```

Comment:
Please use [`Array.prototype.filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) instead of rolling your own version